tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -922,6 +935,289 @@ HRESULT DoSimpleReWrite(_In_ DxcLangExtensionsHelper *pHelper,
+    const bool IgnoreSysRoot = true;
+    if (dxcutil::IsAbsoluteOrCurDirRelative(A->getValue())) {
+      HSOpts.AddPath(A->getValue(), frontend::Angled, IsFrameworkFalse,
+                     IgnoreSysRoot);
+    } else {
+      std::string s("./");
+      s += A->getValue();
+      HSOpts.AddPath(s, frontend::Angled, IsFrameworkFalse, IgnoreSysRoot);
+    }
+  }
+
+  if (msfPtr) {
+    msfPtr->SetupForCompilerInstance(compiler);
+  }
+
+  if (pDefines) {
+    std::string newDefines = compiler.getPreprocessor().getPredefines();
+    newDefines += pDefines;
+    compiler.getPreprocessor().setPredefines(newDefines);
+  }
+}
+
+void PreprocessResult(CompilerInstance &compiler, LPCSTR pFileName) {
+  // These settings are back-compatible with fxc.
+  clang::PreprocessorOutputOptions &PPOutOpts =
+      compiler.getPreprocessorOutputOpts();
+  PPOutOpts.ShowCPP = 1;           // Print normal preprocessed output.
+  PPOutOpts.ShowComments = 0;      // Show comments.
+  PPOutOpts.ShowLineMarkers = 1;   // Show \#line markers.
+  PPOutOpts.UseLineDirectives = 1; // Use \#line instead of GCC-style \# N.
+  PPOutOpts.ShowMacroComments = 0; // Show comments, even in macros.
+  PPOutOpts.ShowMacros = 0;        // Print macro definitions.
+  PPOutOpts.RewriteIncludes = 0;   // Preprocess include directives only.
+
+  FrontendInputFile file(pFileName, IK_HLSL);
+  clang::PrintPreprocessedAction action;
+  if (action.BeginSourceFile(compiler, file)) {
+    action.Execute();
+    action.EndSourceFile();
+  }
+}
+
+class RewriteVisitor : public RecursiveASTVisitor<RewriteVisitor> {
+public:
+  RewriteVisitor(Rewriter &R, TranslationUnitDecl *tu, RewriteHelper &helper)
+      : TheRewriter(R), SourceMgr(R.getSourceMgr()), tu(tu),
+        helper(helper), bNeedLineInfo(false) {}
+
+  bool VisitFunctionDecl(FunctionDecl *f) {
+    if (helper.unusedFunctions.count(f)) {
+      bNeedLineInfo = true;
+
+      TheRewriter.RemoveText(f->getSourceRange());
+      return true;
+    }
+
+    AddLineInfoIfNeed(f->getLocStart());
+    return true;
+  }
+
+  bool VisitTypeDecl(TypeDecl *t) {
+    if (helper.unusedTypes.count(t)) {
+      bNeedLineInfo = true;
+      TheRewriter.RemoveText(t->getSourceRange());
+      return true;
+    }
+    AddLineInfoIfNeed(t->getLocStart());
+    return true;
+  }
+
+  bool VisitVarDecl(VarDecl *vd) {
+    if (vd->getDeclContext() == tu) {
+      if (helper.unusedGlobals.count(vd)) {
+        bNeedLineInfo = true;
+        TheRewriter.RemoveText(vd->getSourceRange());
+        return true;
+      }
+
+      AddLineInfoIfNeed(vd->getLocStart());
+    }
+    return true;
+  }
+
+private:
+  void AddLineInfoIfNeed(SourceLocation Loc) {
+    if (bNeedLineInfo) {
+      bNeedLineInfo = false;
+      auto lineStr = MakeLineInfo(Loc);
+      TheRewriter.InsertTextBefore(Loc, lineStr);
+    }
+  }
+  std::string MakeLineInfo(SourceLocation Loc) {
+    if (Loc.isInvalid())
+      return "";
+    if (!Loc.isFileID())
+      return "";
```

Comment:
Not sure how this would happen, but a comment explaining that line information is needed here, but unavailable might be helpful.

Reply:
I'd like to see when this happens, but suspect we don't want it to be noisy here.  Builtins don't have locations, though I don't think we should be printing those normally.